src/stirling/source_connectors/socket_tracer/protocols/common/event_parser.h

```diff
@@ -120,28 +120,32 @@ ParseResult ParseFrames(message_type_t type, DataStreamBuffer* data_stream_buffe
     buf.remove_prefix(start_pos);
   }
 
-  // Grab size before we start, so we know where the new parsed frames are.
-  const size_t prev_size = frames->size();
-
   // Parse and append new frames to the frames vector.
-  ParseResult result = ParseFramesLoop(type, buf, frames, state);
+  std::deque<TFrameType> new_frames = std::deque<TFrameType>();
+  ParseResult result = ParseFramesLoop(type, buf, &new_frames, state);
 
-  VLOG(1) << absl::Substitute("Parsed $0 new frames", frames->size() - prev_size);
+  VLOG(1) << absl::Substitute("Parsed $0 new frames", new_frames.size());
 
   // Match timestamps with the parsed frames.
   for (size_t i = 0; i < result.frame_positions.size(); ++i) {
     auto& f = result.frame_positions[i];
     f.start += start_pos;
     f.end += start_pos;
 
-    auto& msg = (*frames)[prev_size + i];
+    auto& msg = new_frames[i];
     StatusOr<uint64_t> timestamp_ns_status =
         data_stream_buffer->GetTimestamp(data_stream_buffer->position() + f.end);
     LOG_IF(ERROR, !timestamp_ns_status.ok()) << timestamp_ns_status.ToString();
     msg.timestamp_ns = timestamp_ns_status.ValueOr(0);
   }
   result.end_position += start_pos;
 
+  // Parse frames into map
+  for (auto& frame : new_frames) {
+    // GetStreamID returns 0 by default if not implemented in protocol.
+    TKey key = GetStreamID<TKey, TFrameType>(&frame);
+    (*frames)[key].push_back(frame);
+  }
```

Comment:
Have you tried pushing the `GetStreamID` call into the `ParseFramesLoop` function? It seems like we could avoid this loop if GetStreamID could be called [here](https://github.com/pixie-io/pixie/blob/782674dec0e3a6924b7a3c3002abe13bae10fdd0/src/stirling/source_connectors/socket_tracer/protocols/common/event_parser.h#L225) as the frame is pushed into the deque.

Reply:
I originally had it in `ParseFramesLoop`, but this required changing how `ParseResult` is used to update the timestamp of each frame in lines 130-141. The code became rather verbose and required changing various test files that use `ParseFramesLoop` like `src/stirling/source_connectors/socket_tracer/protocols/{cql,dns,kafka,mysql}/parse_test.cc`, so I opted to move it into `ParseFrames` instead.

Happy to discuss further if you think it would be better to move it back into ParseFramesLoop.

Reply:
On second thought, I think it is better to have the call to `GetStreamID` in `ParseFramesLoop`. To avoid this PR getting too large, I've moved that change to #1732